File: plugins/plugin-elizacloud/typescript/models/tokenization.ts
import type {
  DetokenizeTextParams,
  IAgentRuntime,
  ModelTypeName,
  TokenizeTextParams,
} from "@elizaos/core";
import { ModelType } from "@elizaos/core";
import { encodingForModel, type TiktokenModel } from "js-tiktoken";

type ProcessEnvLike = Record<string, string | undefined>;

function getProcessEnv(): ProcessEnvLike {
  if (typeof process === "undefined") {
    return {};
  }
  return process.env as ProcessEnvLike;
}

const env = getProcessEnv();

async function tokenizeText(
  model: ModelTypeName,
  prompt: string,
): Promise<number[]> {
  const modelName =
    model === ModelType.TEXT_SMALL
      ? (env.ELIZAOS_CLOUD_SMALL_MODEL ?? env.SMALL_MODEL ?? "gpt-5-nano")
      : (env.LARGE_MODEL ?? "gpt-5-mini");
  const tokens = encodingForModel(modelName as TiktokenModel).encode(prompt);
  return tokens;
}

async function detokenizeText(
  model: ModelTypeName,
  tokens: number[],
): Promise<string> {
  const modelName =
    model === ModelType.TEXT_SMALL
      ? (env.ELIZAOS_CLOUD_SMALL_MODEL ?? env.SMALL_MODEL ?? "gpt-5-nano")
      : (env.ELIZAOS_CLOUD_LARGE_MODEL ?? env.LARGE_MODEL ?? "gpt-5-mini");
  return encodingForModel(modelName as TiktokenModel).decode(tokens);
}

export async function handleTokenizerEncode(
  _runtime: IAgentRuntime,
  { prompt, modelType = ModelType.TEXT_LARGE }: TokenizeTextParams,
): Promise<number[]> {
  return await tokenizeText(modelType ?? ModelType.TEXT_LARGE, prompt);
}

export async function handleTokenizerDecode(
  _runtime: IAgentRuntime,
  { tokens, modelType = ModelType.TEXT_LARGE }: DetokenizeTextParams,
): Promise<string> {
  return await detokenizeText(modelType ?? ModelType.TEXT_LARGE, tokens);
}

File: plugins/plugin-elizacloud/rust/src/models/tokenization.rs
#![allow(missing_docs)]

use crate::error::{ElizaCloudError, Result};
use crate::types::{DetokenizeTextParams, ElizaCloudConfig, TokenizeTextParams};

fn get_model_name(config: &ElizaCloudConfig, model_type: &str) -> String {
    if model_type == "TEXT_SMALL" {
        config.small_model.clone()
    } else {
        config.large_model.clone()
    }
}

pub async fn handle_tokenizer_encode(
    config: ElizaCloudConfig,
    params: TokenizeTextParams,
) -> Result<Vec<u32>> {
    let model_name = get_model_name(&config, &params.model_type);

    let bpe = tiktoken_rs::get_bpe_from_model(&model_name)
        .or_else(|_| tiktoken_rs::cl100k_base())
        .map_err(|e| ElizaCloudError::Configuration(format!("Failed to get tokenizer: {}", e)))?;

    let tokens = bpe.encode_ordinary(&params.prompt);
    Ok(tokens.into_iter().collect())
}

pub async fn handle_tokenizer_decode(
    config: ElizaCloudConfig,
    params: DetokenizeTextParams,
) -> Result<String> {
    let model_name = get_model_name(&config, &params.model_type);

    let bpe = tiktoken_rs::get_bpe_from_model(&model_name)
        .or_else(|_| tiktoken_rs::cl100k_base())
        .map_err(|e| ElizaCloudError::Configuration(format!("Failed to get tokenizer: {}", e)))?;

    let text = bpe
        .decode(params.tokens)
        .map_err(|e| ElizaCloudError::Configuration(format!("Failed to decode tokens: {}", e)))?;

    Ok(text)
}
